lab_2_tokenize_by_bpe/main.py

```diff
@@ -26,53 +40,147 @@ def collect_frequencies(
+                elif token == pair[0] and word[index + 1] == pair[1]:
+                    new_word.append(pair[0] + pair[1])
+
+            value = word_freq_updated.pop(word)
+            word_freq_updated[tuple(new_word)] = value
+
+    return word_freq_updated
 
 
 def train(
-    word_frequencies: dict[tuple[str, ...], int] | None, num_merges: int
+        word_frequencies: dict[tuple[str, ...], int] | None, num_merges: int
 ) -> dict[tuple[str, ...], int] | None:
     """
     Creates required number of new tokens by merging existing ones
     :param word_frequencies: dictionary of a kind <preprocessed word: number of occurrences>
     :param num_merges: required number of new tokens
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(num_merges, int):
+        return None
+
+    pair_dict = count_tokens_pairs(word_frequencies)
+    if pair_dict is None:
+        return None
+
+    for _ in range(num_merges):
+        if pair_dict != {}:
+            max_value = max(pair_dict.values())
+            pair_list = [key for key, value in pair_dict.items() if value == max_value]
+
+            pair_list.sort()
+            pair_to_merge = max(pair_list, key=lambda x: len(x[0] + x[1]))
+
+            if isinstance(word_frequencies, dict):
+                word_frequencies = merge_tokens(word_frequencies, pair_to_merge)
+
+            pair_dict.pop(pair_to_merge)
+            if isinstance(word_frequencies, dict):
+                pair_dict = count_tokens_pairs(word_frequencies)
+            if pair_dict is None:
+                return None
+
+    return word_frequencies
 
 
 def get_vocabulary(
-    word_frequencies: dict[tuple[str, ...], int], unknown_token: str
+        word_frequencies: dict[tuple[str, ...], int], unknown_token: str
 ) -> dict[str, int] | None:
     """
     Establishes correspondence between tokens and its integer identifier
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :param unknown_token: a token to signify an unknown token
     :return: dictionary in the form of <token: identifier>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(unknown_token, str):
+        return None
+
+    tokens_list = [unknown_token]
+    for word in word_frequencies:
+        tokens_list.append(word[0])
+        for token in word[0]:
+            if token not in tokens_list:
+                tokens_list.append(token)
+
+    tokens_list.sort()
+    tokens_list.sort(key=len, reverse=True)
+
+    tokens_dict = {}
+    for index, token in enumerate(tokens_list):
+        tokens_dict[token] = index
+
+    return tokens_dict
 
 
 def decode(
-    encoded_text: list[int] | None, vocabulary: dict[str, int] | None, end_of_word_token: str | None
+        encoded_text: list[int] | None, vocabulary: dict[str, int] | None
+        , end_of_word_token: str | None
```

Comment:
there must be an easier way for this func

Reply:
fixed?